### ISSUE TITLE
HTML: Avoid linebreaks inside end tags

### ISSUE THREAD

**Issue by lydell** [prettier/prettier#8302]
Due to the unfortunate whitespace rules of HTML, Prettier has to do some clever tricks to be able to safely format HTML as good as it can while not introducing unwanted spacing in the rendered output of the HTML.

Here’s a tweak we can do: Many times we don’t gain anything by putting a line break inside an end tag. For example:

```html
<div id="update-available">
  Myapp has a new version.
  <a href="#" onClick="window.location.reload(true);return false;"
    >Click to reload</a
  >
</div>
```

That would be much nicer formatted as:

```html
<div id="update-available">
  Myapp has a new version.
  <a href="#" onClick="window.location.reload(true);return false;"
    >Click to reload</a>
</div>
```

But it is unfortunately not as easy as just _never_ putting a linebreak inside end tags, because sometimes that’s the only way to be able to break long lines:

```html
<div id="update-available">
  Myapp has a new version.
  <a href="#" onClick="reload()"
    ><strong
      >Click to reload and get the latest and greatest version!</strong
    ></a
  ><span data-icon="star"></span>
</div>
```

If we were to never break inside end tags, it would look like this:

```html
<div id="update-available">
  Myapp has a new version.
  <a href="#" onClick="reload()"
    ><strong
      >Click to reload and get the latest and greatest version!</strong></a><span data-icon="star"></span>
</div>
```

That line is getting very long.

So my proposal is:

**If an end tag is followed by whitespace, don’t break inside the end tag.**

---

An alternative rule would be:

**Only break inside end tags if immediately followed by another tag or a comment.**

For example, this:

```html
<p id="v1">
  That was done using
  <a href="https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver"
    >MutationObserver</a
  >s and other cool DOM APIs such as
  <a href="TODO-link">IntersectionObserver</a>s. Cool, huh?
</p>

<p id="v2">
  That was done using
  <a href="https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver"
    >MutationObserver</a
  >s and other cool DOM APIs such as
  <a
    href="https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver"
    >IntersectionObserver</a
  >s. Cool, huh?
</p>
```

…could be formatted as:

```html
<p id="v1">
  That was done using
  <a href="https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver"
    >MutationObserver</a>s and other cool DOM APIs such as
  <a href="TODO-link">IntersectionObserver</a>s. Cool, huh?
</p>

<p id="v2">
  That was done using
  <a href="https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver"
    >MutationObserver</a>s and other cool DOM APIs such as
  <a
    href="https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver"
    >IntersectionObserver</a>s. Cool, huh?
</p>
```

**Comment by mkantor** [comment]
I noticed some surprising formatting today:

```html
<pre>
a
b</pre>
```

Becomes:

```html
<pre>
a
b</pre
>
```

[Playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeADgJzgPgDpQCGBARqgPRa4EgA0IE6MAltAM7KiGaYQDuABW4IOKQgDcIzACZ0QJTITABrODADK6JcygBzZDEwBXOPTgBbEnGnTrAGUJ6jhXXABiETOcIwWe5CCERjAQcgAWMOYANgDqYczwbFpgcOoiCcziCQCeAWBsHPQ6bHCYMAKKut7IAGaEUSX0AFZsAB4AQooqauqE5nB2OnC19Y0gLa3qOrpRcACKRhDwIw2mIFqYJZgBjrpGUdxyWDowMTIwYcgAHAAM9FgQJTGK6AFUW+LD9ACOi-AVjFEgTYAFooHBrNY5NhfsxsBUXNUkHVVvQSuZmAZjGs2NNZgslsNkaM1jBCCQztILsgACz0QyEZhRaYAYQg5iRICg0C+ICMJQAKuTRCixuITABJKC2WDqMCYZhMACC0vUMGysxWJQAvtqgA)

---

I can't tell if that's within the scope of this issue. The outcome is a linebreak inside of an end tag, but the surprising part isn't where the linebreak _is_, it's that Prettier is trying to break up that line _at all_ (since it's way under the print width). 

Please let me know if I should file a separate issue!

**Comment by lydell** [comment]
@maxloh It sounds like you’re looking for the [HTML Whitespace Sensitivity](https://prettier.io/docs/en/options.html#html-whitespace-sensitivity) option. That’s not what this issue is about.

**Comment by davidbielik** [comment]
This fixed it for me (in `.prettierrc` or wherever you manage your prettier settings):

```
  "htmlWhitespaceSensitivity": "ignore"
```

**Comment by AliChraghi** [comment]
the comments is't offtopic. last comment solved it for me

**Comment by superclarkk** [comment]
I think the following is related to this bug. Would someone kindly confirm?

Before Prettier:
```
    <form action="" method="post"><label for="">First Name</label><input type="text" name="firstName" id="" value=""><button type="submit">Submit</button></form>
```

After Prettier:
```
    <form action="" method="post">
      <label for="">First Name</label
      ><input type="text" name="firstName" id="" value="" /><button
        type="submit"
      >
        Submit
      </button>
    </form>
```

Ideal formatting (or, at the very least, not inserting new lines in the middle of tags):
```
    <form action="" method="post">
      <label for="">First Name</label>
      <input type="text" name="firstName" id="" value="" />
      <button type="submit">Submit</button>
    </form>
```

**Comment by jerrygreen** [comment]
100% agree on the author above. This looks so much better (non-prettier):

```html
    <form action="" method="post">
      <label for="">First Name</label>
      <input type="text" name="firstName" id="" value="" />
      <button type="submit">Submit</button>
    </form>
```

Rather than this (prettier):

```html
    <form action="" method="post">
      <label for="">First Name</label
      ><input type="text" name="firstName" id="" value="" /><button
        type="submit"
      >
        Submit
      </button>
    </form>
```

I'm having exact same problem.

**Comment by scholtz** [comment]
I dont get it .. so many people find this ugly and there is no solution for 3 years? 

i suggest there will be configuration parameter allowLineBreaksAtEndTags: true|false and if it is set to true, it will not allow end tag to have new space, but instead it place the end tag as whole on new line

**Comment by Foxy-null** [comment]
Still no fix even though we are already in 2024 :(

**Comment by jerrygreen** [comment]
Make a fork of Prettier, where you get rid of the issue, release an installable cli, without the need to build sources, make an extension for vscode, – and I’m gonna probably install it 😂

Too many people want to use it. Too few people want to make it. And the last ones are actually important to solving this issue. Not the wanting ones.

This issue would probably have much more thumbs up, if it was so obvious to see. It can appear only on rare occasions. I’ve noticed this issue only after a year or two of usage somehow. Not sure how it comes up I didn’t see it earlier.

Personally, I don’t think tags should be separated at all. I.e. `<a>` should never ever be separated. Either opening one or closing one. So `</a>` shouldn’t be separated either.

**Comment by Turbulentarius** [comment]
It's not just end-tags though. I am a bit of a purist perfectionist on this admittedly, and previously I would manually format my HTML. One line break+two spaces per block element usually.

Unfortunately I now live with this issue to save time and hopefully enjoy some uniformity in our dev environments. But, it is a huge annoyance that I occasionally spend time and energy wondering about and researching on. I wonder specifically if this is based on some agreed upon standard? It just looks super odd.

In my opinion long lines is not really an issue at all in HTML. Line breaks and spaces are welcome where white space already exists naturally in the actual text content, since it just collapses, but never inside opening or closing tags unless there's a bad amount of properties (which is a code-smell). In these horrible classitis times, I guess it's much to wish for...

I may be the exception, but I hate looking at HTML attributes that's formatted over multiple lines, and often it's only like a few attributes **that can easily fit**. The only time I want breaks here, is when there are numerous of attributes, and then we are at least talking like 10+ attributes (something unusual for well-written HTML). I don't think there should, ever, be a linebreak after less than "<" or before greater than ">" — not sure there ever would be a reason for it.

**Comment by CorySchulz** [comment]
Unfortunately I agree that disabling prettier completely for .html files is the best solution.

**Comment by tedbarth** [comment]
Or use `<!-- prettier-ignore -->` to stop prettier for specific lines.

But I really hope we can get an option to specify per (HTML) element tag name whether introduction of line-breaks is fine or not.
We found the need for this for `<textarea>` (and our own custom elements), where adding line-breaks (around and in content) will make a semantic difference. Formatting should never affect the semantics. [Playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeeAPGBDATnbAPmxNLPIsquptpICNUB6THfIgHShABoQIADjACW0AM7JQeXBADuABTwIJKbADcIwgCa8Q9XNjABrODADK2ALZwAMsKhxkAM2wAbMXD4R6AKzhgYAHUDAWQQAXwPXDVHPn1DE3MBQ3sAc2QYXABXTxAPS2EM7Ny4DAE4XGFrWDcAeXKDGAhceQgxYRFoMIQdPlKGqoQcVwAVCqg8YTgVF3dc9qhU1zgARSyIeGc3Dz4fMQwzNOW1jcckWZ2QAEd1+HkZARUQbDEAWgc4LU-dTOxhVzSAGEIJZLNgwlBoLE8kc4ABBGCZYT0LJ3Cp2BxbOZ8AAWMEsrkCOI602SYDgZmUHWEag6AE8wmAxBI+GocgBJKBfWBmMCVIRwrlmGB05ZYy4RNpwYLYUIoCLTCoxXT2KIwe7YVJg8W5ZK4KJhbCLLKuPC6CL2ILaGA45AADgADHx8DdhPgNVrwedtrkcPRAtbbUgACx8LIeEbYegzH19Sz0T5fLQ2I2pLKauAAMWaYMRaUNqIguj2GAAQgZjKYLNYMWcLnAAL4NoA)

**Comment by slavnycoder** [comment]
Same issue

**Comment by wischi-chr** [comment]
I just saw that and was certain I misconfigured something. Shouldn't prettier make things prettier and not uglier 🤐?`

**Comment by camiblanch** [comment]
I was able to use `"htmlWhitespaceSensitivity": "ignore"` to get the lines to format correctly. I had to add that setting, undo the bad wrap formatting, then save the file again and it formatted correctly

**Comment by wischi-chr** [comment]
So I really did miss a setting. I tried it and it worked for most cases, so now I do a step before Prettier where I remove all line breaks and then run Prettier, and it looks very much like what I expected. Thank you.

Update: Turns out removing all the line breaks is also a stupid idea 🤦‍♂️ so I'll probably stick with `"htmlWhitespaceSensitivity": "ignore"` and just live with all those cases where it doesn't get me the results I hoped for.

**Comment by camiblanch** [comment]
@wischi-chr Okay so it mostly works. After using it today and testing it across a handful of files, I am seeing this small caveat on how prettier is handling formatting anchor tags. With `"htmlWhitespaceSensitivity": "ignore"` it ends up moving periods outside of anchor tags to new lines, which in my case ends up rendering as a space which is not great. 

For example:

```html
<p>
	Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc vestibulum bibendum orci, sed
	porttitor eros iaculis et. Maecenas mollis, purus non aliquet laoreet, dolor nisl placerat nibh,
	ac consectetur metus
	<a href="https://www.github.com">GitHub</a>
	.
</p>
``` 

Renders as

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc vestibulum bibendum orci, sed porttitor eros iaculis et. Maecenas mollis, purus non aliquet laoreet, dolor nisl placerat nibh, ac consectetur metus [GitHub](https://www.github.com/) .
